### PR TITLE
ARROW-5892 : [C++][Gandiva] Support function aliases

### PR DESCRIPTION
**Pull request by pprudhvi**
Access precompiled functions with different names.

### CODE REVIEW

File: c_glib/gandiva-glib/native-function.cpp [hunk 1]
```diff
@@ -145,7 +145,7 @@ ggandiva_native_function_to_string(GGandivaNativeFunction *native_function)
 {
   auto gandiva_native_function =
     ggandiva_native_function_get_raw(native_function);
-  auto gandiva_function_signature = gandiva_native_function->signature();
+  auto gandiva_function_signature = gandiva_native_function->signatures().back();
```
Comment (praveenbingo) [inline c_glib/gandiva-glib/native-function.cpp]: wont they missing the aliases now if we do this..

File: cpp/src/gandiva/expression_registry.h [hunk 2]
```diff
@@ -36,22 +36,26 @@ class FunctionRegistry;
 /// data types and functions supported by Gandiva.
 class GANDIVA_EXPORT ExpressionRegistry {
  public:
-  using iterator = const NativeFunction*;
+  using nf_iterator = const NativeFunction*;
```
Comment (praveenbingo) [inline cpp/src/gandiva/expression_registry.h]: better names please..its hard to follow otherwise..

**Review by praveenbingo — COMMENTED** [review]
please also add a test from gandiva java..

File: cpp/src/gandiva/expression_registry.h (hunk 2 above)
Comment (pprudhvi) [reply cpp/src/gandiva/expression_registry.h]: renamed nf_iterator to native_func_iterator_type and fs_iterator to func_sig_iterator_type.
thanks for reviewing

File: cpp/src/gandiva/function_registry_test.cc [hunk 3]
```diff
@@ -31,7 +31,8 @@ TEST_F(TestFunctionRegistry, TestFound) {
 
   const NativeFunction* function = registry_.LookupSignature(add_i32_i32);
   EXPECT_NE(function, nullptr);
-  EXPECT_EQ(function->signature(), add_i32_i32);
+  EXPECT_TRUE(std::find(function->signatures().begin(), function->signatures().end(),
```
Comment (fsaintjacques) [inline cpp/src/gandiva/function_registry_test.cc]: This should do the trick `EXPECT_THAT(add_i32_i32, In(function->signatures())`.

**Comment by kou** [comment]
I'll push a commit for GLib related part to this branch in a few days.

File: cpp/src/gandiva/native_function.h [hunk 4]
```diff
@@ -45,24 +45,30 @@ class GANDIVA_EXPORT NativeFunction {
...
+                 const DataTypeVector& param_types, DataTypePtr ret_type,
+                 const ResultNullableType& result_nullable_type,
                  const std::string& pc_name, int32_t flags = 0)
-      : signature_(base_name, param_types, ret_type),
+      : signatures_(),
         flags_(flags),
         result_nullable_type_(result_nullable_type),
-        pc_name_(pc_name) {}
+        pc_name_(pc_name) {
+    aliases.insert(aliases.begin(), base_name);
```
[the remark this answers is not in the document]
Comment (pprudhvi) [reply cpp/src/gandiva/native_function.h]: done. thanks for reviewing.

File: cpp/src/gandiva/function_registry_test.cc (hunk 3 above)
Comment (pprudhvi) [reply cpp/src/gandiva/function_registry_test.cc]: using `Contains` matcher from gmock. couldn't find `In`.

File: cpp/src/gandiva/function_registry_common.h [hunk 5]
```diff
@@ -74,31 +74,31 @@ typedef std::unordered_map<const FunctionSignature*, const NativeFunction*, KeyH
 // - NULL handling is of type NULL_IF_NULL
 //
 // The pre-compiled fn name includes the base name & input type names. eg. add_int32_int32
-#define BINARY_SYMMETRIC_SAFE_NULL_IF_NULL(NAME, TYPE)                             \
-  NativeFunction(#NAME, DataTypeVector{TYPE(), TYPE()}, TYPE(), kResultNullIfNull, \
-                 ARROW_STRINGIFY(NAME##_##TYPE##_##TYPE))
+#define BINARY_SYMMETRIC_SAFE_NULL_IF_NULL(NAME, ALIASES, TYPE)          \
+  NativeFunction(#NAME, ALIASES, DataTypeVector{TYPE(), TYPE()}, TYPE(), \
```
Comment (bkietz) [inline cpp/src/gandiva/function_registry_common.h]: Macros which take initializer_lists are tricky because the lack of a paren means that an initializer_list with more than one element is considered two tokens. For example:

```c++
#define VECTOR(T, elements) ::std::vector<T>(elements)
VECTOR(int, {1, 2, 3}); // error: the macro's arguments are "int", " {1", " 2", " 3}"
```

If you add the container type explicitly then it should work to just enclose the initializer_list in parens:
```suggestion
  NativeFunction(#NAME, 
std::vector<std::string> ALIASES, DataTypeVector{TYPE(), TYPE()}, TYPE(), \
```

Please add a function with at least two aliases (in a test would be acceptable) to demonstrate how the macro can be used with plural aliases.

**Review by bkietz — CHANGES_REQUESTED** [review]

File: cpp/src/gandiva/function_registry_common.h (hunk 5 above)
Comment (pprudhvi) [reply cpp/src/gandiva/function_registry_common.h]: changed. Thanks for reviewing.

Comment (bkietz) [reply cpp/src/gandiva/function_registry_common.h]: To clarify: you don't need to wrap *all* initializer lists in parentheses, just those with more than a single element

**Review by pravindra — APPROVED** [review]
